### PR TITLE
Add defauting of OperatingSystemSpec

### LINKED ISSUE #1028: InvalidConfiguration Error after upgrade of KubeOne
After I upgraded from some earlier version (not sure which exactly) to the the recent kubone `1.2.3` used machine contorller `docker.io/kubermatic/machine-controller:v1.33.0` I get on before applied Machines (by Machine Deploymenet) the error on deletion:
```
An error of type = InvalidConfiguration, with message = Failed to
    parse MachineSpec: operatingSystemSpec in the MachineDeployment cannot be empty
    occurred. Please manually delete machine-delete-finalizer finalizer from the machine
    object.
```
It seams in earlier versions it was ok to not have the `operatingSystemSpec` set, now it leads to this error.
The `machine` object:
```
apiVersion: cluster.k8s.io/v1alpha1
kind: Machine
metadata:
  creationTimestamp: "2021-05-16T15:48:34Z"
  deletionGracePeriodSeconds: 0
  deletionTimestamp: "2021-07-28T22:14:56Z"
  finalizers:
  - machine-delete-finalizer
  - machine-node-delete-finalizer
  generateName: run-alpha-az-b-5b86b45956-
  generation: 4744
  labels:
    machine-template-hash: "1642601512"
    machinepool: run-alpha-az-b
  name: run-alpha-az-b-5b86b45956-jhzl4
  namespace: kube-system
  ownerReferences:
  - apiVersion: cluster.k8s.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: MachineSet
    name: run-alpha-az-b-5b86b45956
    uid: 43fa1642-3a5e-4b9a-8deb-d46c6885b8f6
  resourceVersion: "348576566"
  uid: a2b080f3-ae5a-4925-9005-eeca6342dcad
spec:
  metadata:
    creationTimestamp: null
    labels:
      machinepool: run-alpha-az-b
    name: run-alpha-az-b-5b86b45956-jhzl4
  providerSpec:
    value:
      cloudProvider: gce
      cloudProviderSpec:
        assignPublicIPAddress: true
        diskSize: 100
        diskType: pd-ssd
        labels:
          machinepool: run-alpha-az-b
        machineType: n1-standard-2
        network: https://www.googleapis.com/compute/v1/projects/kubermatic-run/global/networks/run-2-alpha
        preemptible: false
        regional: true
        subnetwork: https://www.googleapis.com/compute/v1/projects/kubermatic-run/regions/europe-west1/subnetworks/run-2-alpha-subnet
        tags:
        - firewall
        - targets
        - run-alpha-az-b
        zone: europe-west1-b
      labels:
        machinepool: run-alpha-az-b
      operatingSystem: ubuntu
      sshPublicKeys:
      - |...
  versions:
    kubelet: 1.20.7
status:
  addresses:
  - address: 35.195.10.42
    type: ExternalIP
  - address: run-alpha-az-b-5b86b45956-jhzl4.europe-west1-b.c.kubermatic-run.internal
    type: InternalDNS
  - address: run-alpha-az-b-5b86b45956-jhzl4.c.kubermatic-run.internal
    type: InternalDNS
  - address: run-alpha-az-b-5b86b45956-jhzl4
    type: InternalDNS
  - address: 10.240.0.239
    type: InternalIP
  conditions:
  - lastHeartbeatTime: null
    lastTransitionTime: null
    status: "True"
    type: Ready
  errorMessage: 'An error of type = InvalidConfiguration, with message = Failed to
    parse MachineSpec: operatingSystemSpec in the MachineDeployment cannot be empty
    occurred. Please manually delete machine-delete-finalizer finalizer from the machine
    object.'
  errorReason: DeleteError
  nodeRef:
    apiVersion: v1
    kind: Node
    name: run-alpha-az-b-5b86b45956-jhzl4
    resourceVersion: "345167616"
    uid: f5bc58e2-7130-4b05-be73-2fd44b7400d5
  versions:
    kubelet: v1.20.7
```

### PR DESCRIPTION
**Pull request by kron4eg**
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged)*:
Fixes #1028

**Optional Release Note**:

```release-note
Add defauting of OperatingSystemSpec
```

### CODE REVIEW

**Review by moadqassem — COMMENTED** [review]
@kron4eg Can you please add a small briefing about the purpose of this PR? I see in the issue description that, in the past we allowed having os spec without an os. I would say this was a bug and it got solved. Now, how this PR is pursuing the workaround fix that was suggested in the ticket?

**Comment by moadqassem** [comment]
> @kron4eg Can you please add a small briefing about the purpose of this PR? I see in the issue description that, in the past we allowed having os spec without an os. I would say this was a bug and it got solved. Now, how this PR is pursuing the workaround fix that was suggested in the ticket?

Ah ok, after checking the machine deployment attached in the ticket I know what's the issue now.

File: pkg/providerconfig/types.go [hunk 1]
```diff
@@ -147,3 +155,22 @@ func NewConfigVarResolver(ctx context.Context, client ctrlruntimeclient.Client)
...
+
+func DefaultOperatingSystemSpec(osys providerconfigtypes.OperatingSystem, operatingSystemSpec runtime.RawExtension) (runtime.RawExtension, error) {
+	switch osys {
+	case providerconfigtypes.OperatingSystemAmazonLinux2:
+		return amzn2.DefaultConfig(operatingSystemSpec), nil
+	case providerconfigtypes.OperatingSystemCentOS:
+		return centos.DefaultConfig(operatingSystemSpec), nil
+	case providerconfigtypes.OperatingSystemFlatcar:
+		return flatcar.DefaultConfig(operatingSystemSpec), nil
+	case providerconfigtypes.OperatingSystemRHEL:
```
Comment (moadqassem) [inline pkg/providerconfig/types.go:167]: This actually won't work for RHEL, the reason behind that is, the operating system spec holds all the needed information to register the os such as the username, password and offline token. Without those it won't be possible to get a running reel machine. I would suggest to throw an error when reel is used

Comment (kron4eg) [reply pkg/providerconfig/types.go:167]: Yeah, good point

Comment (kron4eg) [reply pkg/providerconfig/types.go:166]: done!

**Review by moadqassem — COMMENTED** [review]
@kron4eg I guess you need to adjust the tests and how they are running for RHEL. RHEL tests criteria are now different than the others which means you need to do some selection on what should run for RHEL.

**Comment by kron4eg** [comment]
@moadqassem last try before start changing tests 😅

**Comment by moadqassem** [comment]
> @moadqassem last try before start changing tests 😅

I am not sure it will pass at some point, but who knows :D 
```
provider_test.go:262: error getting userdata: 'failed to parse OperatingSystemSpec: unexpected end of JSON input'
```

File: pkg/providerconfig/types.go (hunk 1 above)
Comment (kron4eg) [reply pkg/providerconfig/types.go:167]: >Without those it won't be possible to get a running rhel machine.
Turns out, on some clouds that's not the case (on AWS you could perfectly well have RHEL machines without registration). So the logic is restored to the original one.

**Review by moadqassem — APPROVED** [review]
/approve
/lgtm

**Pull request merged**